### PR TITLE
Wip

### PR DESCRIPTION
**Pull request by TalBenHamouMoveo**
non-bonus tasks completed.

### CODE REVIEW

File: src/app/pokemon.service.spec.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import { TestBed } from '@angular/core/testing';
```
Comment (noyflaysher) [inline src/app/pokemon.service.spec.ts]: remove this file.

File: src/app/app.component.html [hunk 2]
```diff
@@ -0,0 +1,6 @@
+<app-header [title]="title"></app-header>
```
Comment (noyflaysher) [inline src/app/app.component.html]: You don't need to move the title as an input. Write the title in the header component

File: src/app/app.component.html [hunk 3]
```diff
@@ -0,0 +1,6 @@
...
+<div class="row">
```
Comment (noyflaysher) [inline src/app/app.component.html]: remove the classes you don't use

File: src/app/app.component.html [hunk 4]
```diff
@@ -0,0 +1,6 @@
...
+    <div class="container">
```
Comment (noyflaysher) [inline src/app/app.component.html]: The same

File: src/app/components/header/header.component.scss [hunk 5]
```diff
@@ -0,0 +1,37 @@
+.navbar {
```
Comment (noyflaysher) [inline src/app/components/header/header.component.scss:1]: Remove comments

File: src/app/components/header/header.component.scss [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  background-color: #cbc758; // Navbar background color
```
Comment (noyflaysher) [inline src/app/components/header/header.component.scss]: Use variable for the color

File: src/app/components/header/header.component.scss [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+  color: #fff; // Text color
```
Comment (noyflaysher) [inline src/app/components/header/header.component.scss]: The same

File: src/app/components/header/header.component.scss [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    margin: 0;
+    padding: 0;
+
+    .list-item {
+      display: inline-block;
+      margin-left: 20px; // Adjust spacing between items
+      font-size: 22px;
+
+      a {
+        color: #fff;
```
Comment (noyflaysher) [inline src/app/components/header/header.component.scss]: Use variable

File: src/app/components/header/header.component.scss [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+        text-decoration: none;
+
+        &:hover {
+          color: #ffcc00; // Change color on hover
+        }
```
Comment (noyflaysher) [inline src/app/components/header/header.component.scss:33]: Use variable

File: src/app/pages/home/home.component.html [hunk 10]
```diff
@@ -0,0 +1,3 @@
+<div class="row">
```
Comment (noyflaysher) [inline src/app/pages/home/home.component.html]: remove the class if you don't use

File: src/app/components/pokemon-list/pokemon-list.component.ts [hunk 11]
```diff
@@ -0,0 +1,12 @@
+import { Component, Input } from '@angular/core';
+import { Pokemon } from '../../shared/pokemon';
+
+@Component({
+  selector: 'app-pokemon-list',
+  templateUrl: './pokemon-list.component.html',
+  styleUrl: './pokemon-list.component.scss'
+})
+export class PokemonListComponent {
+  @Input() pokemons_server: Pokemon[];
```
Comment (noyflaysher) [inline src/app/components/pokemon-list/pokemon-list.component.ts]: Change the name of the input.

File: src/app/components/pokemon-list/pokemon-list-item/pokemon-list-item.component.ts [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+export class PokemonListItemComponent implements OnInit{
+  @Input() pokemon: Pokemon;
+
+  onClick(pokemon: Pokemon) {
+    this.pokemonService.setClickedPokemon(pokemon);
+  }
+
+  constructor(private pokemonService: PokemonService) { }
+  
+  ngOnInit(): void {
```
Comment (noyflaysher) [inline src/app/components/pokemon-list/pokemon-list-item/pokemon-list-item.component.ts]: Remove oninit if ypu don't use it

File: src/app/components/pokemon-list/pokemon-list-item/pokemon-list-item.component.html [hunk 13]
```diff
@@ -0,0 +1,10 @@
+<div class="container">
+    <a routerLink="details/{{pokemon.name}}" (click)="onClick(pokemon)" class="card-link">
+        <div class="card">
+            <img src="{{pokemon.image}}" style="height: 150px; width: 150px;">
```
Comment (noyflaysher) [inline src/app/components/pokemon-list/pokemon-list-item/pokemon-list-item.component.html]: move the style to scss file

File: src/app/shared/pokemon.ts [hunk 14]
```diff
@@ -0,0 +1,9 @@
+export class Pokemon {
```
Comment (noyflaysher) [inline src/app/shared/pokemon.ts]: Move this file to a type folder

File: src/app/shared/pokemon.service.ts [hunk 15]
```diff
@@ -0,0 +1,70 @@
+import { Injectable } from '@angular/core';
+import { Pokemon } from './pokemon';
+import { HttpClient } from '@angular/common/http';
+import { Observable, forkJoin, map, mergeMap } from 'rxjs';
```
Comment (noyflaysher) [inline src/app/shared/pokemon.service.ts]: move this file to service folder

File: src/app/shared/pokemon.service.ts [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+
+
+const URL = 'https://pokeapi.co/api/v2/pokemon/?offset=0&limit=100/';
+  
+@Injectable({
+  providedIn: 'root'
+})
+export class PokemonService {
+  pokemons: Pokemon[];
+  clickedPokemon: Pokemon;
```
Comment (noyflaysher) [inline src/app/shared/pokemon.service.ts]: You can to know what is the selected pokemon by the param in route. You don't need to save it here

File: src/app/shared/pokemon.service.ts [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+  setClickedPokemon(pokemon: Pokemon) {
+    this.clickedPokemon = pokemon;
+  }
+
+  fetchPokemons(): Observable<any> {
+    return this.http.get<any>(URL);
+  }
+
+  getPokemonsFromUrl(): Pokemon[] {
+    this.fetchPokemons().subscribe((response: any) => {
```
Comment (noyflaysher) [inline src/app/shared/pokemon.service.ts]: implement fetch inside this function. You don't need to separate

File: src/app/shared/pokemon.service.ts [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+      this.extractPokemonData(response.results).subscribe((result) => {
```
Comment (noyflaysher) [inline src/app/shared/pokemon.service.ts]: Add type to result

File: src/app/shared/pokemon.service.ts [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+    });
+    return this.pokemons;
+  }
+
+  getPokemonDetails(url: string): Observable<any> {
+    return this.http.get<any>(url);
+  }
+
+  extractPokemonData(results: any[]): Observable<any[]> {
+    // Create an array of observables for getPokemonTypes
```
Comment (noyflaysher) [inline src/app/shared/pokemon.service.ts]: remove comments

File: src/app/shared/pokemon.service.ts [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+                image: `https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/${result.url.split('/')[6]}.png`,
+                type: typesArray[index] // Assign the resolved types array
+            }));
+        })
+    );
+  }
+
+  getPokemonTypes(url: string): Observable<string[]> {
+      return this.http.get<any>(url).pipe(
+          map((response: any) => {
```
Comment (noyflaysher) [inline src/app/shared/pokemon.service.ts]: add type

File: src/app/components/pokemon-list/pokemon-list-item/pokemon-list-item.component.scss [hunk 21]
```diff
@@ -0,0 +1,37 @@
+.container {
```
Comment (noyflaysher) [inline src/app/components/pokemon-list/pokemon-list-item/pokemon-list-item.component.scss:3]: Use nesting

File: src/app/components/pokemon-list/pokemon-list-item/pokemon-list-item.component.scss [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+.card-link {
+    text-decoration: none; // Remove underline from the link
+    color: inherit; // Inherit the text color from the parent
+    width: calc(33.33% - 40px); // Adjust width as per your requirement
+    margin: 30px; // Adjust margin as per your requirement
+
+    .card {
+        background-color: #ff1f1f;
```
Comment (noyflaysher) [inline src/app/components/pokemon-list/pokemon-list-item/pokemon-list-item.component.scss]: use variables for color

File: src/app/components/pokemon-list/pokemon-list-item/pokemon-list-item.component.scss [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+        border: 1px solid #ccc;
```
Comment (noyflaysher) [inline src/app/components/pokemon-list/pokemon-list-item/pokemon-list-item.component.scss]: use variables for color

File: src/app/components/pokemon-list/pokemon-list-item/pokemon-list-item.component.scss [hunk 24]
```diff
@@ -0,0 +1,37 @@
...
+        display: flex;
+        justify-content:center;
+        align-items:center;
+
+        &:hover {
+            transform: translateY(-5px);
+            box-shadow: 10px 10px 10px rgba(0, 0, 0, 1);
+        }
+
+        font-family: 'Ariel', sans-serif; /* Specify your desired font family */
```
Comment (noyflaysher) [inline src/app/components/pokemon-list/pokemon-list-item/pokemon-list-item.component.scss]: use variable for font family

File: src/app/components/pokemon-list/pokemon-list-item/pokemon-list-item.component.scss [hunk 25]
```diff
@@ -0,0 +1,37 @@
...
+        font-size: 20px; /* Adjust font size as needed */
+        font-weight: 400; /* Adjust font weight as needed */
+        color: #ffffff; /* Adjust font color as needed */
```
Comment (noyflaysher) [inline src/app/components/pokemon-list/pokemon-list-item/pokemon-list-item.component.scss]: use variables for color

File: src/app/app.module.ts [hunk 26]
```diff
@@ -0,0 +1,38 @@
...
+import { AppComponent } from './app.component';
+import { PokemonListComponent } from './components/pokemon-list/pokemon-list.component';
+import { PokemonDetailComponent } from './components/pokemon-detail/pokemon-detail.component';
+import { PokemonListItemComponent } from './components/pokemon-list/pokemon-list-item/pokemon-list-item.component';
+import { HeaderComponent } from './components/header/header.component';
+import { RouterModule, Routes } from '@angular/router';
+import { FormsModule } from '@angular/forms';
+import { HomeComponent } from './pages/home/home.component';
+
+const appRoutes: Routes = [
```
Comment (noyflaysher) [inline src/app/app.module.ts]: Extract the routes to the another file - app-routing.module.module.ts

File: src/app/components/pokemon-detail/pokemon-detail.component.scss [hunk 27]
```diff
@@ -0,0 +1,58 @@
+.container {
+    display: flex;
+    flex-direction: column;
+    justify-content: center;
+    align-items: center;
+    text-align: center;
+    background-color: rgb(204, 59, 59);
```
Comment (noyflaysher) [inline src/app/components/pokemon-detail/pokemon-detail.component.scss]: use variables

File: src/app/components/pokemon-detail/pokemon-detail.component.scss [hunk 28]
```diff
@@ -0,0 +1,58 @@
...
+    border: 1px solid #ccc;
```
Comment (noyflaysher) [inline src/app/components/pokemon-detail/pokemon-detail.component.scss]: use variables

File: src/app/components/pokemon-detail/pokemon-detail.component.scss [hunk 29]
```diff
@@ -0,0 +1,58 @@
...
+      width: 300px;
+      height: 300px;
+      margin-bottom: 10px;
+    }
+  
+    label {
+      font-size: 24px;
+      font-weight: bold;
+      margin-bottom: 5px;
+      color: rgb(17, 38, 232);
```
Comment (noyflaysher) [inline src/app/components/pokemon-detail/pokemon-detail.component.scss]: use variables

File: src/app/components/pokemon-detail/pokemon-detail.component.scss [hunk 30]
```diff
@@ -0,0 +1,58 @@
...
+    }
+  }
+  .button {
+    display: inline-block;
+    padding: 10px 20px;
+    font-size: 16px;
+    font-weight: bold;
+    text-decoration: none;
+    text-align: center;
+    border: 2px solid #333;
```
Comment (noyflaysher) [inline src/app/components/pokemon-detail/pokemon-detail.component.scss]: use variables

File: src/app/components/pokemon-detail/pokemon-detail.component.scss [hunk 31]
```diff
@@ -0,0 +1,58 @@
...
+    border-radius: 5px;
+    color: #333;
+    background-color: #fff;
```
Comment (noyflaysher) [inline src/app/components/pokemon-detail/pokemon-detail.component.scss]: use variables

File: src/app/components/pokemon-detail/pokemon-detail.component.scss [hunk 32]
```diff
@@ -0,0 +1,58 @@
...
+  .button {
+    display: inline-block;
+    padding: 10px 20px;
+    font-size: 16px;
+    font-weight: bold;
+    text-decoration: none;
+    text-align: center;
+    border: 2px solid #333;
+    border-radius: 5px;
+    color: #333;
```
Comment (noyflaysher) [inline src/app/components/pokemon-detail/pokemon-detail.component.scss]: use variables

File: src/app/components/pokemon-detail/pokemon-detail.component.scss [hunk 33]
```diff
@@ -0,0 +1,58 @@
...
+    background-color: #fff;
+    cursor: pointer;
+    transition: background-color 0.3s, color 0.3s, border-color 0.3s;
+  
+    &:hover {
```
Comment (noyflaysher) [inline src/app/components/pokemon-detail/pokemon-detail.component.scss:50]: use variables

File: src/main.ts
Comment (noyflaysher) [inline src/main.ts:1]: - use mixins and nesting
- use variables for font and colors
- remove unused classes
- change to meaningful names

File: src/app/components/pokemon-list/pokemon-list-item/pokemon-list-item.component.html [hunk 34]
```diff
@@ -0,0 +1,10 @@
+<div class="container">
+    <a routerLink="details/{{pokemon.name}}" (click)="onClick(pokemon)" class="card-link">
```
Comment (noyflaysher) [inline src/app/components/pokemon-list/pokemon-list-item/pokemon-list-item.component.html]: Send the id not the name

File: src/app/components/pokemon-detail/pokemon-detail.component.ts [hunk 35]
```diff
@@ -0,0 +1,29 @@
...
+})
+export class PokemonDetailComponent implements OnInit{
+  pokemon: Pokemon;
+  id: string;
+
+  constructor(private route: ActivatedRoute, private pokemonService: PokemonService, private router: Router) { }
+
+  ngOnInit(): void {
+    this.id = this.route.snapshot.params['id'];
+    if (this.pokemonService.getPokemonsList() == undefined)
```
Comment (noyflaysher) [inline src/app/components/pokemon-detail/pokemon-detail.component.ts]: ===

File: src/app/components/pokemon-detail/pokemon-detail.component.ts [hunk 36]
```diff
@@ -0,0 +1,29 @@
...
+  id: string;
+
+  constructor(private route: ActivatedRoute, private pokemonService: PokemonService, private router: Router) { }
+
+  ngOnInit(): void {
+    this.id = this.route.snapshot.params['id'];
+    if (this.pokemonService.getPokemonsList() == undefined) 
+      this.pokemonService.getPokemonsFromUrl();
+
+    this.pokemon = this.pokemonService.getPokemonsList().filter(pokemon => pokemon.id === this.id).at(0);
```
Comment (noyflaysher) [inline src/app/components/pokemon-detail/pokemon-detail.component.ts]: this.pokemonService.getPokemonsList().filter(pokemon => pokemon.id === this.id).at(0)

enter this to function

File: src/app/components/pokemon-list/pokemon-list-item/pokemon-list-item.component.scss [hunk 37]
```diff
@@ -0,0 +1,43 @@
+@import '../../../../styles.scss';
```
Comment (noyflaysher) [inline src/app/components/pokemon-list/pokemon-list-item/pokemon-list-item.component.scss:1]: try to remove style.scss from all the scss files

File: src/app/service/pokemon.service.ts [hunk 38]
```diff
@@ -0,0 +1,74 @@
...
+
+
+const URL = 'https://pokeapi.co/api/v2/pokemon/?offset=0&limit=10/';
+  
+@Injectable({
+  providedIn: 'root'
+})
+export class PokemonService {
+  pokemons: Pokemon[];
+  clickedPokemon: Pokemon;
```
Comment (noyflaysher) [inline src/app/service/pokemon.service.ts]: remove clicked pokemon

File: src/app/service/pokemon.service.ts [hunk 39]
```diff
@@ -0,0 +1,74 @@
...
+
+  constructor(private http: HttpClient) { }
+
+  getClickedPokemon(): Pokemon {
+    return this.clickedPokemon;
+  }
+  setClickedPokemon(pokemon: Pokemon) {
```
Comment (noyflaysher) [inline src/app/service/pokemon.service.ts]: remove

File: src/app/components/pokemon-list/pokemon-list-item/pokemon-list-item.component.ts [hunk 40]
```diff
@@ -0,0 +1,18 @@
...
+@Component({
+  selector: 'app-pokemon-list-item',
+  templateUrl: './pokemon-list-item.component.html',
+  styleUrl: './pokemon-list-item.component.scss'
+})
+export class PokemonListItemComponent {
+  @Input() pokemon: Pokemon;
+
+  onClick(pokemon: Pokemon) {
+    this.pokemonService.setClickedPokemon(pokemon);
```
Comment (noyflaysher) [inline src/app/components/pokemon-list/pokemon-list-item/pokemon-list-item.component.ts]: remove onClick

File: src/app/components/pokemon-list/pokemon-list-item/pokemon-list-item.component.scss (hunk 37 above)
Comment (TalBenHamouMoveo) [reply src/app/components/pokemon-list/pokemon-list-item/pokemon-list-item.component.scss:1]: its must

**Pull request merged**